src/Conversion/ONNXToKrnl/Tensor/Unique.cpp

```diff
@@ -0,0 +1,194 @@
+      axis = axis < 0 ? axis + rank : axis;
+      assert(axis >= 0 && axis < rank && "axis is out of bound");
+    }
+    //
+    // Emit a Unique call to get the outputs to count unique
+    // slices(subtensors) of X along axis at first.
+    //
+    Type indexTy = rewriter.getIndexType();
+    Value iZero = create.math.constantIndex(0);
+    Value uniqueCount = create.mem.alloca(MemRefType::get({}, indexTy));
+    create.krnl.store(iZero, uniqueCount, {});
+    Value noneValue;
+    emitArgUnique(rewriter, loc, uniqueCount, X, axis, /*sorted=*/sorted,
+        noneValue, noneValue, noneValue, noneValue, /*count_only=*/true);
+    //
+    // Calculate shapes of output Tensors
+    //
+    Value total = create.krnl.load(uniqueCount, {});
+    NonAffineIndexExpr totalDimExpr = DimIndexExpr(total);
+    DimsExpr outputYDims;
+    DimsExpr outputIndexDims;
+    DimsExpr outputInverseIndexDims;
+    if (axis < 0) {
+      outputYDims.emplace_back(totalDimExpr);
+      outputIndexDims.emplace_back(totalDimExpr);
+      DimIndexExpr inputDimExpr = LiteralIndexExpr(xShape[0]);
+      for (int64_t i = 1; i < rank; i++) {
+        inputDimExpr = inputDimExpr * LiteralIndexExpr(xShape[i]);
+      }
+      outputInverseIndexDims.emplace_back(inputDimExpr);
+    } else {
+      for (int64_t i = 0; i < rank; i++) {
+        DimIndexExpr tDimExpr = LiteralIndexExpr(xShape[i]);
+        if (i == axis)
+          tDimExpr = totalDimExpr;
+        outputYDims.emplace_back(tDimExpr);
+      }
+      outputIndexDims.emplace_back(totalDimExpr);
+      outputInverseIndexDims.emplace_back(LiteralIndexExpr(xShape[axis]));
+    }
+    //
+    // Insert an allocation and deallocation for the outputs.
+    //
+    Value outputY;
+    if (axis < 0) {
+      MemRefType memrefType =
+          MemRefType::get({ShapedType::kDynamic}, elementType);
+      outputY = create.mem.alignedAlloc(memrefType, outputYDims);
+    } else {
+      ArrayRef<int64_t> xShape = getShape(X.getType());
+      SmallVector<int64_t> yShape;
+      for (int i = 0; i < rank; i++)
+        yShape.emplace_back((i == axis) ? ShapedType::kDynamic : xShape[i]);
+      MemRefType memrefType = MemRefType::get(yShape, elementType);
+      outputY = create.mem.alignedAlloc(memrefType, outputYDims);
+    }
+    Type i64Type = rewriter.getI64Type();
+    MemRefType memrefType = MemRefType::get({ShapedType::kDynamic}, i64Type);
+    Value indices = create.mem.alignedAlloc(memrefType, outputIndexDims);
+    Value inverse_indices =
+        create.mem.alignedAlloc(memrefType, outputInverseIndexDims);
+    Value counts = create.mem.alignedAlloc(memrefType, outputIndexDims);
+    //
+    // Emit a Unique call to get the outputs
+    //
+    create.krnl.store(iZero, uniqueCount, {});
+    emitArgUnique(rewriter, loc, uniqueCount, X, axis, /*sorted=*/sorted,
+        outputY, indices, inverse_indices, counts, /*count_only=*/false);
+    rewriter.replaceOp(op, {outputY, indices, inverse_indices, counts});
```

Comment:
Some of the outputs are optional. How do you handle that? Need to check the getResults().size()?

Reply:
Types of optional outputs becomes noneType if they are not required, passed as NULL pointers in OMTensorUnique,
and OMTensorUnique prepares necessary outputs only.